gpii/node_modules/gpii-oauth2/gpii-oauth2-datastore/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,629 @@
+            clientId: clientRecord.client.id
+        };
+        var authDecisionPromise = gpii.oauth2.dbDataStore.findRecord(findAuthDecisionByGpiiTokenAndClientIdDataSource, directModel);
+
+        var mapper = function (authDecision) {
+            return authDecision ? {
+                accessToken: authDecision.accessToken
+            } : undefined;
+        };
+        var mappedPromise = fluid.promise.map(authDecisionPromise, mapper);
+        fluid.promise.follow(mappedPromise, promiseTogo);
+    }
+    return promiseTogo;
+
+};
+// ==== End of findAccessTokenByOAuth2ClientIdAndGpiiToken()
+
+// The post data process function for implementing findAuthorizedClientsByGpiiToken()
+gpii.oauth2.dbDataStore.findAuthorizedClientsByGpiiTokenPostProcess = function (data) {
+    var records = [];
+    fluid.each(data, function (row) {
+        var oneResult = {
+            authDecisionId: row.id,
+            oauth2ClientId: row.doc.oauth2ClientId,
+            clientName: row.doc.name,
+            selectedPreferences: row.value.selectedPreferences
+        };
+        records.push(oneResult);
+    });
+    return records;
+};
+
+// The post data process function for implementing findAuthByAccessToken()
+gpii.oauth2.dbDataStore.findAuthByAccessTokenPostProcess = function (data) {
+    var result = {
+        userGpiiToken: data.value.gpiiToken,
+        selectedPreferences: data.value.selectedPreferences,
+        oauth2ClientId: data.doc.oauth2ClientId
+    };
+    return result;
+};
+
+// Authorization Codes
+// -------------------
+
+gpii.oauth2.dbDataStore.saveAuthCode = function (saveDataSource, authDecisionId, code) {
+    var promiseTogo = fluid.promise();
+    var data = {
+        authDecisionId: authDecisionId,
+        code: code
+    };
+
+    var emptyFields = gpii.oauth2.dbDataStore.filterEmptyFields(data, ["authDecisionId", "code"]);
+
+    if (emptyFields.length > 0) {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingInput, {fieldName: emptyFields.join(" & ")});
+        promiseTogo.reject(error);
+    } else {
+        promiseTogo = gpii.oauth2.dbDataStore.addRecord(saveDataSource, gpii.oauth2.dbDataStore.docTypes.authCode, "id", data);
+    }
+
+    return promiseTogo;
+};
+
+// The post data process function for implementing findAuthByCode()
+gpii.oauth2.dbDataStore.findAuthByCodePostProcess = function (data) {
+    if (data.doc.revoked === false) {
+        var result = {
+            clientId: data.doc.clientId,
+            redirectUri: data.doc.redirectUri,
+            accessToken: data.doc.accessToken
+        };
+        return result;
+    } else {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.unauthorizedAuthCode, {code: data.id});
+        return error;
+    }
+};
+
+// Client Credentials Tokens
+// -------------------------
+
+gpii.oauth2.dbDataStore.addClientCredentialsToken = function (saveDataSource, clientCredentialsTokenData) {
```

Comment:
Docs needed here - especially for the peculiar "emptiness" semantics for clientCredentialsTokenData. Why might it be an empty object?